### ISSUE TITLE
Shouldn't we be checking all files in the repo?

### ISSUE THREAD

**Issue by 2bndy5** [shenxianpeng/cpp-linter-action#25]
This occurred to me when I was adding the `diff-only` option. If `diff-only` is false, then shouldn't we be examining all source files in the repo that use one of the designated file `extensions`? Currently, we are only examining files that have been changed despite what `diff-only` option is set to.

However, we can only do this if the repo is checked out before this action is run, but there's the concern of clever people also checking out their repo's submodules (which shouldn't be examined by this action). I suppose we can use the REST API to list all files in the repo, but the REST API's responses are limited to 30 entries per page. If the repo has more than 30 files, then we'd have to subsequently keep making REST requests until we have a list of all the files in the repo. It would definitely be easier if the repo was checked out beforehand, and we use python to "walk" through the repo's files to examine any applicable source files.

**Comment by shenxianpeng** [comment]
It’s best to help users check or output prompts to tell them that they need to have `uses: actions/checkout@v2` to checkout repo before running cpp-linter-action. Otherwise, they may also encounter the same problem as me in #24.

If their code has a submodule or subtree, they should also check out if they need it. cpp-linter-action only focuses on the work of clang-format and clang-tidy.

So if the repo was checked out beforehand, there is no need to download the source code through the REST API, right?

**Comment by 2bndy5** [comment]
> It’s best to help users check or output prompts to tell them that they need to have uses: actions/checkout@v2 to checkout repo before running cpp-linter-action.

I did alter the example yml in the readme with a comment about this. I do like the idea of having something in the log about non-existent files. Currently, all we're logging about this is
```
INFO:CPP Linter:Downloading file from url: https://github.com/shenxianpeng/test-cpp-linter-action/raw/fd6ec9c55cf93df9e01dad042f80e7835fc2fa5f/demo.cpp
```
but it should say more to make the user aware that the action is trying to recover from no git checkout.

>So if the repo was checked out beforehand, there is no need to download the source code through the REST API, right?

True, but it doesn't hurt too much to leave it in as a recourse.

---
>If their code has a submodule or subtree, they should also check out if they need it. cpp-linter-action only focuses on the work of clang-format and clang-tidy.

Checking out submodules is often done for complex projects' build CI. Furthermore, the compiler database file that clang-tidy looks for is generated at build time (ei using `cmake -bbuild .`). If we do look in the user's repo for all source files using one of the specified extensions, then this action will inadvertantly crawl any submodule(s) that were also checked out. If this issue becomes a desired feature, then I'd use the _.gitmodules_ file (something generated and preferably maintained by git bash commands) to ignore any source files in the checked out submodules. Did I understand your comment correctly?

**Comment by 2bndy5** [comment]
<details><summary>example of a <i>.gitmodules</i> file</summary>

```ini
[submodule "RF24"]
	path = RF24
	url = https://github.com/nRF24/RF24.git
[submodule "RF24Network"]
	path = RF24Network
	url = https://github.com/nRF24/RF24Network.git
[submodule "RF24Mesh"]
	path = RF24Mesh
	url = https://github.com/nRF24/RF24Mesh.git
[submodule "pybind11"]
	path = pybind11
	url = https://github.com/pybind/pybind11.git
```
</details>

The paths are relative to the repo's root (which can be specified using the input option `repo-root`).

**Comment by 2bndy5** [comment]
I added a conditional warning in the log about missing files when no git checkout was used.

**Comment by shenxianpeng** [comment]
> If their code has a submodule or subtree, they should also check out if they need it. cpp-linter-action only focuses on the work of clang-format and clang-tidy.

I mean we don't need to consider submodules, users should consider it by themselves. The working process of cpp-linter-action is to have a complete code, then to analyze the specified code changes and post results comments.

I don't know if I misunderstand your comment about submodules.

**Comment by shenxianpeng** [comment]
> I added a conditional warning in the log about missing files when no git checkout was used.

Tested and saw the warning looks good in [log](https://pipelines.actions.githubusercontent.com/KJSIOnyFZcI4wN2LGl5dZSPe7IwMuTqWkMeL99FDIxcBTatsZ3/_apis/pipelines/1/runs/42/signedlogcontent/3?urlExpires=2021-10-05T12%3A56%3A32.8839085Z&urlSigningMethod=HMACV1&urlSignature=xWOwa19jO3sRNh3O%2BhX7g7QpF8buJUSTT%2BKXzKvwqKs%3D)

**Comment by 2bndy5** [comment]
Yeah I think there's a misunderstanding. I'm trying to say the same thing you are: Don't pay attention to submodules. Although I keep going into detail about how to identify if a path in the repo is a submodule (using _.gitmodules_).

The important thing I'm questioning is: 
### Should we examine all source files or only the files that changed?
Currently, we are only examining the files that changed.
- set the `diff-only` option to "true" if examining only changed files
- set the `diff-only` option to "false" if examining all applicable source files

I think I'll start another branch on my fork to see what this looks like.

**Comment by shenxianpeng** [comment]
I think I see. most CI checks are only checking the changed files, such as SonarQube Quality Gates. I would like to set `diff-only` option to "true" by default.
In my case examining only changed files is good, if the codebase needs format I'd like to do a bulk change commits before, then introduce an action for future changed files examination.

**Comment by 2bndy5** [comment]
The way I have it setup now (in master-py branch): If we set the `diff-only` option to "true" it will only checks the lines (of the changed files) shown in the diff. Do you still want to check the entire file if diff-only is true?

**Comment by shenxianpeng** [comment]
Oh... `diff-only` refers to lines, not files. 
I tested diff-only to "true" in the test-cpp-linter-action repo, compared with `diff-only` is "false", I don't see any different output, maybe the test code is not enough. so I'm not sure if need to changed `diff-only` option to "true", I follow your opinion.

**Comment by 2bndy5** [comment]
The `diff-only` option was meant as a entrypoint to posting comments directly on the line in the file (#11), but it may not be as useful as I thought. The difference between lines and files does seem confusing. Maybe I should rename that to `changed-lines-only` (or `diff-lines-only`) instead. Then, I could add an option called `changed-files-only`. Can you think of better names? (I seem terrible with naming things).

**Comment by shenxianpeng** [comment]
In GitHub, the diff files button called "Files changed", maybe we name them `files-changed-only` and `lines-changed-only`? 
Your naming has always been very professional, and my pleasure to accept any changes.

**Comment by 2bndy5** [comment]
I like your rationality. I'm going with your name ideas. TBC...

**Comment by shenxianpeng** [comment]
Great discovery, log output command looks really good 💯

**Comment by 2bndy5** [comment]
I'm close to submitting a PR, but I have to refine handling the new `ignore` option. There may be a bug concerning the root folder of the repo. And, I'd like to add a possible syntax that allows excluding an entire folder except for a specific file/folder within the excluded folder.

My current idea:
```yml
      - uses: actions/checkout@v2
      - uses: shenxianpeng/cpp-linter-action@master
        id: linter
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          style: file
          # search entire repo for source files
          file-changed-only: false
          # ignore everything in demo folder except the demo.cpp file. Notice the bang (`!`) prefix
          ignore: |
            # yml comments are not recognized here
            demo 
            !demo/demo.cpp
```

You can find my progress in [this latest run](https://github.com/2bndy5/cpp-linter-action/runs/3821571629?check_suite_focus=true)

**Comment by 2bndy5** [comment]
I tried out using the new log commands to implement in-file comments, but I found some differences compared to using the REST API to post comments.
1. The message's body can't be multiple lines.
2. The [message looks different in the file's diff](https://github.com/2bndy5/cpp-linter-action/commit/7ce1dd61bf0c5c33a2e18e2c0ee8df75695d9eb1) compared to the [message in the workflow's summary](https://github.com/2bndy5/cpp-linter-action/actions/runs/1320584979).
 
However, despite these differences, it is much easier (and even faster) to use log commands.

### What you think?
- Do you want to keep the use of log commands to make comments?
- Do you want to keep the use of REST API to make comments?
- Do you want to use both log commands and REST API to make comments?

**Comment by 2bndy5** [comment]
oh I forgot about the retention period! We shall have both then. PR coming soon...

**Comment by 2bndy5** [comment]
The link to super-linter example is a year old, but the annotations still exist 👍🏼 

>If there are both, it might be better to have a switch to control log command and REST API comment on or off.

This thought had occured to me also. How about a new option called `thread-comment`? If it set true, comment will appear in the thread (PR or commit). Otherwise no comment in the thread.

I also think we should switch away from using a table in the readme to document the input options. They are getting numerous, and the `ignore` option needs its own list to fully explain the capabilities.

**Comment by shenxianpeng** [comment]
Do you mean to call  https://github.com/2bndy5/cpp-linter-action/commit/7ce1dd61bf0c5c33a2e18e2c0ee8df75695d9eb1 as `thread-comment`?, not sure if it is a commit comment, my understand below is a thread comment.
![image](https://user-images.githubusercontent.com/3353385/136663244-37449df9-0423-430d-a29e-f099874a1b69.png)

If there are other better ways to show input options, please do it. Btw, I still like using readme to display the basic usage.
